### PR TITLE
Add shard awareness tests

### PR DESCRIPTION
**Pull request by fruch**
Adapting few tests from scylladb/python-driver, to cover the shard aware.
Building it ontop of #36, to try and test the new usage of shard-aware port (not done yet)

### CODE REVIEW

**Comment by fruch** [comment]
@haaawk a gentle reminder.

**Comment by fruch** [comment]
@haaawk a gentle reminder...

**Review by enaydanov — APPROVED** [review]
Overall it looks good to me, except few magic constants in the introduced code ("100", "10000", "50000") which should have names, imo.

**Comment by fruch** [comment]
> Overall it looks good to me, except few magic constants in the introduced code ("100", "10000", "50000") which should have names, imo.

well the magic there is the knowledge on which shard those keys would land (given SMP=4)  (knowledge that grant to me, only by trying it out on the python-driver)

I guess it should be some kind of mapping, to be clearer

**Comment by bentsi** [comment]
@haaawk Any chance that you can review this?
CC: @slivne

**Comment by haaawk** [comment]
@bentsi This PR is based on not merged PR so I guess we can review it but will that make you any good to have a reviewed PR that can't be merged anyway?

**Comment by haaawk** [comment]
@avelanarius when you have few minutes, please take a look and check whether this PR is in line with the github action integration testing we want to add.

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                "SCYLLA_SHARDING_IGNORE_MSB",
+                new ArrayList<String>() {
+                  {
+                    add("12");
+                  }
+                });
+          }
+        };
+    ShardingInfo.ConnectionShardingInfo sharding = ShardingInfo.parseShardingInfo(params);
+    assertThat(sharding.shardId).isEqualTo(1);
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:76]: Better to use assertEquals so that in case of error the message tells what's the exact mismatch.

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    byte[] byte_array1 = {
+      'a',
+    };
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array1))))
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:81]: ditto

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        .isEqualTo(4);
+
+    byte[] byte_array2 = {
+      'b',
+    };
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array2))))
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:87]: ditto

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        .isEqualTo(6);
+
+    byte[] byte_array3 = {
+      'c',
+    };
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array3))))
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:93]: ditto

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        .isEqualTo(6);
+
+    byte[] byte_array4 = {
+      'e',
+    };
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array4))))
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:99]: ditto

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        .isEqualTo(4);
+
+    byte[] byte_array5 = {'1', '0', '0', '0', '0', '0'};
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array5))))
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:103]: ditto

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                  }
+                });
+          }
+        };
+    ShardingInfo.ConnectionShardingInfo sharding = ShardingInfo.parseShardingInfo(params);
+    assertThat(sharding.shardId).isEqualTo(1);
+
+    byte[] byte_array1 = {
+      'a',
+    };
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:80]: Please make this a single line.

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array1))))
+        .isEqualTo(4);
+
+    byte[] byte_array2 = {
+      'b',
+    };
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:86]: Please make this a single line.

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array2))))
+        .isEqualTo(6);
+
+    byte[] byte_array3 = {
+      'c',
+    };
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:92]: Please make this a single line.

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array3))))
+        .isEqualTo(6);
+
+    byte[] byte_array4 = {
+      'e',
+    };
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:98]: Please make this a single line.

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.testng.annotations.Test;
+
+public class ShardingInfoTest {
+
+  @Test(groups = "unit")
+  public void should_parsing_and_calculating_shard_id() {
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:31]: Convention in Java is to use CamelCase

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    ShardingInfo.ConnectionShardingInfo sharding = ShardingInfo.parseShardingInfo(params);
+    assertThat(sharding.shardId).isEqualTo(1);
+
+    byte[] byte_array1 = {
+      'a',
+    };
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array1))))
+        .isEqualTo(4);
+
+    byte[] byte_array2 = {
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:84]: Convention in Java is to use CamelCase

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+      'b',
+    };
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array2))))
+        .isEqualTo(6);
+
+    byte[] byte_array3 = {
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:90]: Convention in Java is to use CamelCase

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+      'c',
+    };
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array3))))
+        .isEqualTo(6);
+
+    byte[] byte_array4 = {
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:96]: Convention in Java is to use CamelCase

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+      'e',
+    };
+    assertThat(sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(byte_array4))))
+        .isEqualTo(4);
+
+    byte[] byte_array5 = {'1', '0', '0', '0', '0', '0'};
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:102]: Convention in Java is to use CamelCase

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java (hunk 2 above)
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:81]: All those test cases share the same code. Please extract an assert:
```
void assertTargetsShard(byte[] key, int expectedShard) {
  assertEquals(expectedShard, sharding.shardingInfo.shardId(factory.hash(ByteBuffer.wrap(key)))));
}
```

Comment (haaawk) [reply driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:81]: Also each case should be a separate test so that when it fails it's apparent from the JUnit report what the problem is.

**Comment by bentsi** [comment]
@haaawk why the #36 is not merged? Do we want eventually to merge it?

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 16]
```diff
@@ -165,4 +166,36 @@ public void verify_same_shard_in_tracing() throws InterruptedException {
...
+    Cluster cluster =
+        register(
+            Cluster.builder()
+                .addContactPoints(getContactPoints().get(0))
+                .withPort(19042)
+                .withReconnectionPolicy(reconnectionPolicy)
+                .withLoadBalancingPolicy(loadBalancingPolicy)
+                .build());
+    cluster.init();
+    Session session = cluster.connect();
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:195]: Shouldn't cluster and session be closed after the test?

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 17]
```diff
@@ -165,4 +166,36 @@ public void verify_same_shard_in_tracing() throws InterruptedException {
...
+    // this host
+    Cluster cluster =
+        register(
+            Cluster.builder()
+                .addContactPoints(getContactPoints().get(0))
+                .withPort(19042)
+                .withReconnectionPolicy(reconnectionPolicy)
+                .withLoadBalancingPolicy(loadBalancingPolicy)
+                .build());
+    cluster.init();
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:194]: Why is this needed? It does not seem to be necessary https://github.com/datastax/java-driver/tree/3.x/manual#quick-start

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+import java.util.function.Predicate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.Test;
+
+@CreateCCM(PER_CLASS)
+@CCMConfig(numberOfNodes = 3, config = "ring_delay_ms:10000")
+public class ShardAwareTest extends CCMTestsSupport {
+
+  static final Logger logger = LoggerFactory.getLogger(ShardAwareTest.class);
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:36]: Why this is not private?

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+
+  static void createKeyspaceAndColumnFamily(Session session) {
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:38]: Why this is not private?

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+          event.getSourceElapsedMicros(),
+          event.getSource(),
+          event.getThreadName(),
+          event.getDescription());
+      assertThat(event.getThreadName()).isEqualTo(shardName);
+    }
+    assertThat(trace.getEvents().stream().anyMatch(new isLocalQuery<Event>()));
+  }
+
+  void queryData(Session session, boolean verifyInTracing) {
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:99]: Why this is not private?

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+    // this host
+    Cluster cluster =
+        register(
+            Cluster.builder()
+                .addContactPoints(getContactPoints().get(0))
+                .withPort(ccm().getBinaryPort())
+                .withReconnectionPolicy(reconnectionPolicy)
+                .withLoadBalancingPolicy(loadBalancingPolicy)
+                .build());
+    cluster.init();
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:162]: Why is this needed?

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+    Session session = cluster.connect();
+
+    createKeyspaceAndColumnFamily(session);
+    createData(session);
+    this.queryData(session, true);
+  }
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:168]: Shouldn't cluster and session be closed?

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 23]
```diff
@@ -165,4 +166,36 @@ public void verify_same_shard_in_tracing() throws InterruptedException {
...
   }
+
+  @Test(groups = {"short", "scylla"})
+  @CCMConfig(
+      config = {
+        "ring_delay_ms:10000",
+        "native_transport_port:0",
+        "native_shard_aware_transport_port:19042"
+      })
+  public void verify_same_shard_in_tracing_using_only_shard_aware_port()
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:177]: This test is almost identical like the previous one. Why not have a helper method that runs the tests and just uses different port?

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+public class ShardAwareTest extends CCMTestsSupport {
+
+  static final Logger logger = LoggerFactory.getLogger(ShardAwareTest.class);
+
+  static void createKeyspaceAndColumnFamily(Session session) {
+    session.execute("DROP KEYSPACE IF EXISTS preparedtests");
+    session.execute(
+        "CREATE KEYSPACE preparedtests WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '3'}");
+    session.execute("USE preparedtests");
+    session.execute("CREATE TABLE cf0 (a text, b text, c text, PRIMARY KEY (a, b))");
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:43]: In general it would be better to use SchemaBuilder

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+
+  void createData(Session session) {
+    session.execute("USE preparedtests");
+    PreparedStatement prepared = session.prepare("INSERT INTO cf0 (a, b, c) VALUES  (?, ?, ?)");
+
+    BoundStatement bs1 = prepared.bind();
+    bs1.setString("a", "a");
+    bs1.setString("b", "b");
+    bs1.setString("c", "c");
+    session.execute(bs1);
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:54]: It seems that a little helper method would be useful not to duplicate this code 3 times.

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+    session.execute("DROP KEYSPACE IF EXISTS preparedtests");
+    session.execute(
+        "CREATE KEYSPACE preparedtests WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '3'}");
+    session.execute("USE preparedtests");
+    session.execute("CREATE TABLE cf0 (a text, b text, c text, PRIMARY KEY (a, b))");
+  }
+
+  void createData(Session session) {
+    session.execute("USE preparedtests");
+    PreparedStatement prepared = session.prepare("INSERT INTO cf0 (a, b, c) VALUES  (?, ?, ?)");
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:48]: Better to use QueryBuilder

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+    session.execute(bs2);
+
+    BoundStatement bs3 = prepared.bind();
+    bs3.setString("a", "100000");
+    bs3.setString("b", "f");
+    bs3.setString("c", "g");
+    session.execute(bs3);
+  }
+
+  class isLocalQuery<E> implements Predicate<Event> {
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:69]: Could it be private and maybe static?

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java (hunk 11 above)
Comment (avelanarius) [reply driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:31]: Other DataStax tests often use snake_case (although they are not consistent and some use CamelCase).

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+                .withPort(ccm().getBinaryPort())
+                .withReconnectionPolicy(reconnectionPolicy)
+                .withLoadBalancingPolicy(loadBalancingPolicy)
+                .build());
+    cluster.init();
+    Session session = cluster.connect();
+
+    createKeyspaceAndColumnFamily(session);
+    createData(session);
+    this.queryData(session, true);
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:167]: Why `this.`? seems redundant

File: driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java (hunk 7 above)
Comment (avelanarius) [reply driver-core/src/test/java/com/datastax/driver/core/ShardingInfoTest.java:80]: Isn't this formatting that way to keep Maven formatting checker happy?

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 29]
```diff
@@ -0,0 +1,168 @@
...
+          event.getSource(),
+          event.getThreadName(),
+          event.getDescription());
+      assertThat(event.getThreadName()).isEqualTo(shardName);
+    }
+    assertThat(trace.getEvents().stream().anyMatch(new isLocalQuery<Event>()));
+  }
+
+  void queryData(Session session, boolean verifyInTracing) {
+    PreparedStatement prepared = session.prepare("SELECT * FROM cf0 WHERE a=? AND b=?");
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:100]: Better to use QueryBuilder

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 30]
```diff
@@ -165,4 +166,36 @@ public void verify_same_shard_in_tracing() throws InterruptedException {
...
+                .withPort(19042)
+                .withReconnectionPolicy(reconnectionPolicy)
+                .withLoadBalancingPolicy(loadBalancingPolicy)
+                .build());
+    cluster.init();
+    Session session = cluster.connect();
+
+    createKeyspaceAndColumnFamily(session);
+    createData(session);
+    this.queryData(session, true);
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:199]: Drop `this.`

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java (hunk 20 above)
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:99]: verifyInTracing seems to be always true. If so, let's drop the flag.

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 31]
```diff
@@ -0,0 +1,168 @@
...
+    ResultSet results1 = session.execute(bs1);
+    Row row1 = results1.one();
+    assertThat(row1).isNotNull();
+    assertThat(row1.getString("a")).isEqualTo("a");
+    assertThat(row1.getString("b")).isEqualTo("b");
+    assertThat(row1.getString("c")).isEqualTo("c");
+
+    if (verifyInTracing) {
+      this.verifySameShardInTracing(results1, "shard 1");
+    }
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:115]: Seems duplicated 3 times with different params. Let's have a helper function.

File: driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java [hunk 32]
```diff
@@ -0,0 +1,168 @@
...
+    bs1.enableTracing();
+    ResultSet results1 = session.execute(bs1);
+    Row row1 = results1.one();
+    assertThat(row1).isNotNull();
+    assertThat(row1.getString("a")).isEqualTo("a");
+    assertThat(row1.getString("b")).isEqualTo("b");
+    assertThat(row1.getString("c")).isEqualTo("c");
+
+    if (verifyInTracing) {
+      this.verifySameShardInTracing(results1, "shard 1");
```
Comment (haaawk) [inline driver-core/src/test/java/com/datastax/driver/core/ShardAwareTest.java:114]: Let's drop `this.`

**Comment by avelanarius** [comment]
> @avelanarius when you have few minutes, please take a look and check whether this PR is in line with the github action integration testing we want to add.

I think it will even make it easier, as this makes necessary changes to `CcmBridge` to support Scylla. I'd very helpful if @fruch could write what it takes to run those integration tests with Scylla (I think you need to have scylla-ccm installed, some environment variables configured etc. - I haven't run them myself).

**Comment by haaawk** [comment]
> @haaawk why the #36 is not merged? Do we want eventually to merge it?

We do want to merge it but it needs more work and we don't have capacity at the moment. I will try to find someone to do it in the near future.

**Comment by haaawk** [comment]
The PR LGTM. I left some comments but nothing major. The tests seem very nice. Thanks for creating them.

**Comment by bentsi** [comment]
OK. Thanks @haaawk

**Comment by haaawk** [comment]
> > @avelanarius when you have few minutes, please take a look and check whether this PR is in line with the github action integration testing we want to add.
> 
> I think it will even make it easier, as this makes necessary changes to `CcmBridge` to support Scylla. I'd very helpful if @fruch could write what it takes to run those integration tests with Scylla (I think you need to have scylla-ccm installed, some environment variables configured etc. - I haven't run them myself).

Yes. @bentsi @fruch please write in commit message how to run those tests so that everyone knew how to do it

**Comment by avelanarius** [comment]
Progress in this PR: #94 was merged (which contains a small part of this PR).

**Comment by haaawk** [comment]
I think we can unfreeze this PR. Right @avelanarius ?

**Comment by avelanarius** [comment]
> I think we can unfreeze this PR. Right @avelanarius ?

Pushed fixed version as a new PR: #106.
